Selected profile summary download feature

rmathew1011 opened the pull request:
# Description

This PR introduces feature to exporting multiple profile downloads by organization.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing Procedures

Was able to test the same on local set up.

rmathew1011 committed:
WIP - added methods for selected profile summary download

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -46,6 +49,33 @@ public IndividualExportController(@Lazy IndividualRepo repo, @Lazy ExporterRegis
         this.exporterRegistry = exporterRegistry;
     }
 
+    @PostMapping("/individual/{orgId}/export")
```
Please reuse same method as matching path as below. Remove this new method.

Update @GetMapping to @RequestMapping and specify all attributes please, explicitly stating methods POST and GET. The @RequestBody `List<String>` ids can be added as expected but may change to a more generic object encapsulating the export request more specifically. The logic to perform lookup by request body or by id is whether the nullable request body is provided.

```
...
// for POST use request body
List<Individual> individuals = repo.findIndividualsByIds(ids);
...
// for GET use id path parameter
Optional<Individual> individual = repo.findById(id);
...
```

tamu-sad-iii replied:
Please try not to duplicate code, when possible, try to avoid redundant code. Unless a clear extensibility issue presents itself or some constraint requires the duplication.

rmathew1011 replied:
- Removed duplicate method
 - Replaced @GetMapping with @RequestMapping, specifying the multiple methods and single value in it.
 - Incorporated the suggested logic: use request body if provided, else use path variable `id`

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -80,13 +110,13 @@ public IndividualModel process(IndividualModel resource) {
                 addResource(resource, new ResourceLink(
```
Please add a resource for export profiles batch zip request to show in response links.

rmathew1011 replied:
This part of the code was not modified  - there were extra space characters that showed up at end of each lline - I deleted those extra space and hence it showed up as a code change for this PR.

tamu-sad-iii replied:
This part of the code is where to describe the API. These links return in individual responses and some search responses. The UI should be able to use them as an input.

tamu-sad-iii replied:
The additional resource link would be something like "Organization selected profiles export". This is correct right?

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/ZipDocxExporter.java:
```diff
@@ -51,9 +51,72 @@ public String contentDisposition(String filename) {
         return CONTENT_TYPE;
     }
 
+    public StreamingResponseBody streamIndividualsByOrg(List<Individual> individuals, String name) {
```
Please try to follow consistent formatting. Please indent if then code blocks appropriately. Use consistent indentation found in the existing code.

rmathew1011 replied:
Applied formatting based on existing code.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/Exporter.java:
```diff
@@ -27,6 +27,12 @@ public default StreamingResponseBody streamIndividuals(Flux<Individual> cursor,
         ));
     }
 
+    public default StreamingResponseBody streamIndividualsByOrg(List<Individual> individuals, String name) {
```
This method name is not accurate. There is no org provided in the arguments. 

Please rename `streamIndividuals`.

What does this method have to do differently than the `streamIndividual` method below? Is there any redundant code that can be avoided?

rmathew1011 replied:
- Renamed to `streamIndividuals`.
 - I would request keeping both these methods in this interface as they serve different purpose.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/ExporterRegistry.java:
```diff
@@ -20,8 +20,11 @@ public class ExporterRegistry {
             return exporter.get();
         }
+
+        exporters.forEach(e -> System.out.println("Registered exporter: " + e.type()));
```
Please remember to use logging where debug may be useful or remove System.out.

rmathew1011 replied:
Removing this line - had added it to debug.

rmathew1011 committed:
Incorporated PR suggestions

rmathew1011 committed:
Update IndividualExportController.java

Reverted one change - Line 63  - to `type.equals("zip")`

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/config/WebSecurityConfig.java:
```diff
@@ -120,12 +120,17 @@ CorsFilter corsFilter() {
         source.registerCorsConfiguration("/individual/{id}", embedConfig);
         source.registerCorsConfiguration("/individual/search/findByIdIn", embedConfig);
+        source.registerCorsConfiguration("/individual/*/export", embedConfig);
+        source.registerCorsConfiguration("/individual/{id}", embedConfig);
```
Please remove the duplicate.

rmathew1011 replied:
Removed the duplicate entry.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/config/WebSecurityConfig.java:
```diff
@@ -120,12 +120,17 @@ CorsFilter corsFilter() {
+        source.registerCorsConfiguration("/individual/*/export", embedConfig);
+        source.registerCorsConfiguration("/individual/{id}", embedConfig);
+        source.registerCorsConfiguration("/individual/export", embedConfig);
+        source.registerCorsConfiguration("/individual/**", embedConfig);
```
Please remove wild cards in CORS paths.

rmathew1011 replied:
Removed wild cards from CORS path.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/config/WebSecurityConfig.java:
```diff
@@ -218,6 +221,10 @@ protected SecurityFilterChain configure(HttpSecurity http) throws Exception {
+
+                .requestMatchers(POST, "/individual/export").permitAll()
+
+                .requestMatchers(POST, "/individual/*/export").permitAll()
```
Please use the path variable and not a wild card.

Also, I don't think these are even required. Try to remove them and see if it still works as expected.

rmathew1011 replied:
Removed these entries - working as pexpected - as the actual paths specified.

tamu-sad-iii replied:
Please remove them then.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -46,11 +49,15 @@ public IndividualExportController(@Lazy IndividualRepo repo, @Lazy ExporterRegis
     }
 
-    @GetMapping("/individual/{id}/export")
+    @RequestMapping(
```
Please cleanup this method.

tamu-sad-iii replied:
Try returning only once and reuse existing code where possible.

rmathew1011 replied:
Method cleaned up and only one return statement is now present.

tamu-sad-iii replied:
Nicely done.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/ZipDocxExporter.java:
```diff
@@ -51,9 +51,67 @@ public String contentDisposition(String filename) {
+            try (ZipOutputStream zos = new ZipOutputStream(outputStream)) {
+                for (Individual individual : individuals) {
+                    final List<String> type = individual.getType();
+                    Optional<DisplayView> displayView = displayViewRepo.findByTypesIn(type);
```
Please try to create a private method to remove the redundant code added.

rmathew1011 replied:
Created a private method to remove the redundant code added.

rmathew1011 replied:
An additional point  - this line:
`FileOutputStream fos = new FileOutputStream(zipFile.getAbsolutePath());`

`fos` is not being used.

Would you want me to delete this.
if so, please do let me know.
Thank you.

tamu-sad-iii replied:
Unrelated to the current request in this comment thread thus making it unresolvable. If it works without remove it. Or write a file for caching and hash the request body.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -126,7 +146,8 @@ private void addResource(IndividualModel resource, ResourceLink link) {
                 link.getType(),
-                link.getName()
+                link.getName(),
+                null
```
Use an empty array here instead of null please.

rmathew1011 replied:
Added an empty array here instead of null.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/discovery/model/repo/IndividualRepo.java:
```diff
@@ -135,6 +136,18 @@ public List<Individual> findByIdIn(List<String> ids) {
     }
 
+    public List<Individual> findIndividualsByIds(List<String> ids) {
+        if (ids == null || ids.isEmpty()) {
```
This only requires the null check or neither and just don't pass in null.

rmathew1011 replied:
checking for only ids.isEmpty() .

tamu-sad-iii replied:
I said it only requires the null check. And now I take it back. The is empty check prevents a Solr query where no results are expected.

tamu-sad-iii replied:
The guard clause might be more appropriate in the terminal method before Solr query. Thanks for finding this and the unused file output stream.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -60,16 +67,29 @@ public ResponseEntity<StreamingResponseBody> export(
-        if (individual.isPresent()) {
-            Individual document = individual.get();
+        List<Individual> individuals;
+        if( !ids.isEmpty() ) {
```
Please format consistently.

rmathew1011 replied:
Formatting done for this file.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/Exporter.java:
```diff
@@ -27,6 +27,12 @@ public default StreamingResponseBody streamIndividuals(Flux<Individual> cursor,
     }
 
+    public default StreamingResponseBody streamIndividuals(List<Individual> individuals, String name) {
+      throw new UnsupportedExporterTypeException(
```
Please use consistent indentation throughout.

rmathew1011 replied:
Formatting done for this file.

rmathew1011 committed:
Based on the PR suggestions, incorporated the following changes:
1. Removed wild card entries from the cors configuration
2. Removed duplicate entry from the cors configuration
3. Post removal of these wild card entries and specifying the path, the code works.
4. Cleaned up export method to use only one return statement
5. In ZipDocxExporter file - removed the redundant code and incorporated the streaming process and zip process into a private method
6 Code clean up and fixed indentation

rmathew1011 committed:
Fixed indentation and code clean up as per PR review

rmathew1011 committed:
Code clean up and refactor

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/service/ZipDocxExporter.java:
```diff
@@ -51,19 +51,45 @@ public String contentDisposition(String filename) {
+
+            try (
+                FileOutputStream fos = new FileOutputStream(zipFile.getAbsolutePath());
+                ZipOutputStream zos = new ZipOutputStream(outputStream)
```
```suggestion
                ZipOutputStream zos = new ZipOutputStream(outputStream);
```

rmathew1011 committed:
Update src/main/java/edu/tamu/scholars/middleware/export/service/ZipDocxExporter.java

Co-authored-by: tamu-sad-iii <144841721+tamu-sad-iii@users.noreply.github.com>

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java:
```diff
@@ -46,30 +50,57 @@ public IndividualExportController(@Lazy IndividualRepo repo, @Lazy ExporterRegis
         @RequestParam(required = false, defaultValue = "docx") String type,
-        @RequestParam(required = true) String name
+        @RequestParam(required = true) String name,
+        @RequestBody(required = false) List<String> ids
```
Can we make this optional? Is empty check can be ignored as well if the pointer is for enumeration. As long as it is not null.

tamu-sad-iii commented on src/main/java/edu/tamu/scholars/middleware/config/WebSecurityConfig.java:
```diff
@@ -218,6 +219,10 @@ protected SecurityFilterChain configure(HttpSecurity http) throws Exception {
+
+                .requestMatchers(POST, "/individual/export").permitAll()
+
+                .requestMatchers(POST, "/individual/*/export").permitAll()
```
```suggestion
                .requestMatchers(POST, "/individual/{id}/export").permitAll()
```
However, they can be removed.

https://github.com/TAMULib/scholars-discovery/blob/tamu-main/src/main/java/edu/tamu/scholars/middleware/export/controller/IndividualExportController.java#L59C17-L59C101

Admin guard clause in the controller.
